### PR TITLE
Improve the usability of the external services UI

### PR DESCRIPTION
**Pull request by beyang**
Improves the usability of the external services UI with the following changes:
* Declutter the "add external services" page, and remove "external services" in favor of "add repositories".
* Add explicit instructions for each external service and declutter the default config. We previously relied on default config to provide de facto documentation, and this meant a lot of clutter and verbiage in the config box by default.
* Add a "repositories" subsection to the site admin sidebar and remove "external services" in favor of "add/view repositories" language.

The screenshots below provide examples of the above changes.

<table>
<tr>
<th>Before</th><th>After</th>
</tr>
<tr>
<td></td>
<td></td>
</tr>
<tr>
<td></td>
<td></td>
</tr>
<tr>
<td>

</td>
<td>

</td>
</tr>
</table>

This change also makes minor improvements to onboarding and the activation widget, but I'm holding off, because I would like to make larger scale changes to the onboarding experience.

### CODE REVIEW

File: cmd/frontend/graphqlbackend/schema.graphql [hunk 1]
```diff
@@ -818,6 +818,8 @@ input AddExternalServiceInput {
     displayName: String!
     # The JSON configuration of the external service.
     config: String!
+    # Delete the service if validation fails
+    requireValidation: Boolean
```
Comment (tsenart) [inline cmd/frontend/graphqlbackend/schema.graphql]: I can't think of a reason to allow creating an invalid external service, so rather than parameterising it, what do you think of always deleting an external service that failed to be validated?

Comment (keegancsmith) [reply cmd/frontend/graphqlbackend/schema.graphql]: agreed

File: cmd/frontend/graphqlbackend/external_services.go [hunk 2]
```diff
@@ -50,7 +52,16 @@ func (r *schemaResolver) AddExternalService(ctx context.Context, args *struct {
 
 	res := &externalServiceResolver{externalService: externalService}
 	if err := syncExternalService(ctx, externalService); err != nil {
```
Comment (keegancsmith) [inline cmd/frontend/graphqlbackend/external_services.go]: FYI the implementation of `syncExternalService` does two things:
- Validate the configuration correctly speaks to the code host.
- Kick off a sync.

The validation doesn't require the external service to be saved in the DB (in fact I don't think it uses that, we serialize the full external service struct here). I suggest instead of deleting an external service, we do the split and do a call which validates on an unsaved service. WDYT?
We can split that into two, then rather than

File: cmd/frontend/graphqlbackend/site_alerts.go [hunk 3]
```diff
@@ -61,13 +61,6 @@ func (r *siteResolver) Alerts(ctx context.Context) ([]*Alert, error) {
 }
 
 func init() {
-	conf.ContributeWarning(func(c conf.Unified) (problems conf.Problems) {
-		if c.ExternalURL == "" {
-			problems = append(problems, conf.NewSiteProblem("`externalURL` is required to be set for many features of Sourcegraph to work correctly."))
-		}
-		return problems
-	})
```
Comment (keegancsmith) [inline cmd/frontend/graphqlbackend/site_alerts.go]: why is this removed? Some context would be good :) I remember some sort of discussion on this, maybe involving @slimsag?

File: dev/start.sh [hunk 4]
```diff
@@ -73,11 +73,11 @@ export GRAFANA_SERVER_URL=http://localhost:3370
...
-export CRITICAL_CONFIG_FILE=${CRITICAL_CONFIG_FILE:-./dev/critical-config.json}
-export SITE_CONFIG_FILE=${SITE_CONFIG_FILE:-./dev/site-config.json}
-export GLOBAL_SETTINGS_FILE=${GLOBAL_SETTINGS_FILE:-./dev/global-settings.json}
-export SITE_CONFIG_ALLOW_EDITS=true
-export GLOBAL_SETTINGS_ALLOW_EDITS=true
+# export CRITICAL_CONFIG_FILE=${CRITICAL_CONFIG_FILE:-./dev/critical-config.json}
+# export SITE_CONFIG_FILE=${SITE_CONFIG_FILE:-./dev/site-config.json}
+# export GLOBAL_SETTINGS_FILE=${GLOBAL_SETTINGS_FILE:-./dev/global-settings.json}
+# export SITE_CONFIG_ALLOW_EDITS=true
+# export GLOBAL_SETTINGS_ALLOW_EDITS=true
```
Comment (keegancsmith) [inline dev/start.sh]: debug stuff here. Maybe update these to allow setting them to empty variables to avoid start.sh specifying them.

File: enterprise/dev/start.sh [hunk 5]
```diff
@@ -18,13 +18,13 @@ source "$DEV_PRIVATE_PATH/enterprise/dev/env"
...
-export SITE_CONFIG_ALLOW_EDITS=true
-export GLOBAL_SETTINGS_ALLOW_EDITS=true
-export EXTSVC_CONFIG_ALLOW_EDITS=true
+# export CRITICAL_CONFIG_FILE=$DEV_PRIVATE_PATH/enterprise/dev/critical-config.json
+# export SITE_CONFIG_FILE=$DEV_PRIVATE_PATH/enterprise/dev/site-config.json
+# export EXTSVC_CONFIG_FILE=$DEV_PRIVATE_PATH/enterprise/dev/external-services-config.json
+# export GLOBAL_SETTINGS_FILE=$PWD/../dev/global-settings.json
+# export SITE_CONFIG_ALLOW_EDITS=true
+# export GLOBAL_SETTINGS_ALLOW_EDITS=true
+# export EXTSVC_CONFIG_ALLOW_EDITS=true
```
Comment (keegancsmith) [inline enterprise/dev/start.sh]: ditto: more lines you probably don't want to merge.

File: cmd/frontend/graphqlbackend/site_alerts.go (hunk 3 above)
Comment (slimsag) [reply cmd/frontend/graphqlbackend/site_alerts.go]: @beyang I am sure you have some plans here, but we need to push users to set this somehow. I would not suggest merging this PR without having this or something like it in place. Things that will subtly break if `externalURL` is not set correctly include:

- User invites
- Saved searches email notifications
- Basic code intelligence (https://github.com/sourcegraph/infrastructure/issues/1514)
- Extensions (I don't know why)
- Much more

File: web/src/site-admin/SiteAdminExternalServiceForm.tsx [hunk 6]
```diff
@@ -82,7 +86,8 @@ export class SiteAdminExternalServiceForm extends React.Component<Props, {}> {
                     disabled={this.props.loading}
                 >
                     {this.props.loading && <LoadingSpinner className="icon-inline" />}
-                    {this.props.mode === 'edit' ? 'Update external service' : 'Add external service'}
+                    {this.props.submitName ??
+                        (this.props.mode === 'edit' ? 'Update external service' : 'Add external service')}
```
Comment (christinaforney) [inline web/src/site-admin/SiteAdminExternalServiceForm.tsx]: Since we're moving away from "external service" language we should do this consistently for the configuration page too. Not sure if this should be slightly different, other options include: Add/update code host configuration, Configure repositories for <name of code host i.e. GitHub>.

```suggestion
                        (this.props.mode === 'edit' ? 'Update code host' : 'Add code host')}
```

File: web/src/site-admin/SiteAdminExternalServicesPage.tsx [hunk 7]
```diff
@@ -195,19 +195,17 @@ export class SiteAdminExternalServicesPage extends React.PureComponent<Props, St
...
-                <PageTitle title="External services - Admin" />
+                <PageTitle title="Manage repositories - Admin" />
                 <div className="d-flex justify-content-between align-items-center mt-3 mb-3">
-                    <h2 className="mb-0">External services</h2>
+                    <h2 className="mb-0">Manage repositories</h2>
                     <Link
                         className="btn btn-primary e2e-goto-add-external-service-page"
                         to="/site-admin/external-services/new"
                     >
                         <AddIcon className="icon-inline" /> Add external service
```
Comment (christinaforney) [inline web/src/site-admin/SiteAdminExternalServicesPage.tsx]: Should this also be updated to say 'Add repositories'?

File: cmd/frontend/graphqlbackend/schema.graphql (hunk 1 above)
Comment (beyang) [reply cmd/frontend/graphqlbackend/schema.graphql]: I've seen instances on customer calls where an external service gets added, fails validation, but repositories are still successfully added, so I don't have full faith in the validation code. However, I agree about limiting the API surface and this addition isn't actually used by the frontend in the changeset, so I'm just deleting this for now.

File: cmd/frontend/graphqlbackend/site_alerts.go (hunk 3 above)
Comment (beyang) [reply cmd/frontend/graphqlbackend/site_alerts.go]: Added back the `externalURL` warning

File: web/src/site-admin/SiteAdminAddExternalServicesPage.tsx [hunk 8]
```diff
@@ -28,48 +22,45 @@ export class SiteAdminAddExternalServicesPage extends React.Component<Props> {
     /**
      * Gets the external service kind and add-service kind from the URL paramsters
```
Comment (lguychard) [inline web/src/site-admin/SiteAdminAddExternalServicesPage.tsx]:
```suggestion
     * Returns the id of the external service from the URL parameters.
```

This comment seems outdated.

File: web/src/site-admin/SiteAdminAddExternalServicesPage.tsx [hunk 9]
```diff
@@ -28,48 +22,45 @@ export class SiteAdminAddExternalServicesPage extends React.Component<Props> {
...
-                <p>Choose an external service to add to Sourcegraph.</p>
-                {ALL_EXTERNAL_SERVICE_ADD_VARIANTS.map((service, i) => (
-                    <div className="add-external-services-page__card" key={i}>
-                        <ExternalServiceCard to={SiteAdminAddExternalServicesPage.getAddURL(service)} {...service} />
+                <PageTitle title="Add repositories" />
+                <div className="d-flex justify-content-between align-items-center mt-3 mb-3">
+                    <h2 className="mb-0">Add repositories</h2>
+                </div>
+                <p className="mt-2">Add repositories from one of these code hosts.</p>
+                {map(codeHostExternalServices, (externalService, id) => (
```
Comment (felixfbecker) [inline web/src/site-admin/SiteAdminAddExternalServicesPage.tsx]: You can do this without lodash
```suggestion
                {Object.entries(codeHostExternalServices).map(([id, externalService]) => (
```

File: web/src/site-admin/SiteAdminAddExternalServicesPage.tsx [hunk 10]
```diff
@@ -28,48 +22,45 @@ export class SiteAdminAddExternalServicesPage extends React.Component<Props> {
...
+                    <div className="add-external-services-page__card" key={id}>
+                        <ExternalServiceCard to={getAddURL(id)} {...externalService} />
+                    </div>
+                ))}
+                <br />
+                <h2>Other connections</h2>
+                <p className="mt-2">Add connections to non-code-host services.</p>
+                {map(nonCodeHostExternalServices, (externalService, id) => (
```
Comment (felixfbecker) [inline web/src/site-admin/SiteAdminAddExternalServicesPage.tsx]:
```suggestion
                {Object.entries(nonCodeHostExternalServices).map(([id, externalService]) => (
```

File: web/src/site-admin/SiteAdminAddExternalServicesPage.tsx [hunk 11]
```diff
@@ -1,17 +1,11 @@
...
-import {
-    AddExternalServiceMetadata,
-    ALL_EXTERNAL_SERVICE_ADD_VARIANTS,
-    ExternalServiceVariant,
-    getExternalService,
-    isExternalServiceVariant,
-} from './externalServices'
+import { codeHostExternalServices, nonCodeHostExternalServices, allExternalServices } from './externalServices'
 import { SiteAdminAddExternalServicePage } from './SiteAdminAddExternalServicePage'
+import { map } from 'lodash'
```
Comment (felixfbecker) [inline web/src/site-admin/SiteAdminAddExternalServicesPage.tsx]:
```suggestion
```

File: web/src/site-admin/externalServices.tsx [hunk 12]
```diff
@@ -100,32 +101,343 @@ const editorActionComments = {
...
+            >
+                the docs for more options
+            </a>
+            , or try one of the buttons below.
+        </p>
+    </div>
+)
+
+const gitlabEditorActions = (isSelfManaged: boolean): EditorAction[] =>
+    concat(
```
Comment (felixfbecker) [inline web/src/site-admin/externalServices.tsx]: Instead of relying on lodash, you can just spread the arrays into an array literal
